rfd/0135-kube-apps-discovery.md

```diff
@@ -0,0 +1,199 @@
+can be used to specify which protocol to use for the app's URI. 
+
+If Kubernetes service has multiple ports in use, we will expose each port as a separate app, with name of the app including the port's name
+specified in the service. For example, if a service has port 443 exposed with name `tls` and also port `5432` with name `postgres`
+we will create two apps: `testApp-tls-main-cluster-prod`, `testApp-postgres-main-cluster-prod`. Names for ports are mandatory in Kubernetes if more
+than one port is exposed for the service, so we always will be able to use it if we need it. User will be able to use annotation to specify
+preferred port, in that case only app with this port will be created, without port info in the app name.
+
+In order for discovery service to be able to dynamically create/update/delete apps we will add Read/Write permissions 
+for the apps to the built-in system role `Discovery`.
+
+Labels from the Kubernetes service will be copied to the corresponding Teleport app.
+
+## UX
+
+### Discovery
+
+Discovery service will use new `kubernetes` matchers to periodically poll list of services inside the Kubernetes cluster.
+Labels specified in the config will be used to filter out services that should be exposed. Also users will be able to
+specify namespaces to process. Another new field for the configuration of the discovery service will be `kubernetes_cluster` - 
+it specifies kubernetes cluster name in which discovery service is running. It is mandatory if `kubernetes` matchers are
+specified in the config.
+
+```yaml
+## This section configures the Discovery Service
+discovery_service:
+  enabled: yes
+  discovery_group: kube-auto-apps
+  kubernetes_cluster: main-cluster
+  kubernetes:
+    - types: ["app"] # in the future "db" will be possible
+      namespaces: [ "toronto", "porto" ]
+      labels:
+        env: staging
+    - types: ["app"]
+      namespaces: [ "seattle", "oakland" ]
+      labels:
+        env: testing
+```
+
+`kubernetes_cluster` field in the config will be translated into `teleport.dev/kubernetes-cluster` label on the Teleport app resource, so
+later it's be easier to target it in the corresponding app service.
+
+Kubernetes annotation `teleport.dev/discovery-protocol` controls protocol for the access of the Teleport app we create. If annotation is missing
+`tcp` type will be used in the app's URI. Additionally, if kubernetes service port definition has `appProtocol` field, and it contains
+values `http`/`https` we will use it in the URI, but annotation supersedes hint from the `appProtocol`.
+
+Annotation `teleport.dev/discovery-port` will control preferred port for the Kubernetes service. Its value should be one of the
+exposed service ports; otherwise, the app will not be imported. Value can be matched either by numeric value or by the name of
+the port defined on the service.
+
+Example Kubernetes service configs using these annotations can look like this:
+```yaml
```

Comment:
Would there be any harm if we import all services by default using some criteria? This will help users to discover the feature and opt out if they are not interested.

@r0mant @marcoandredinis @kimlisa @AntonAM for example, in Discover we can ask users if they want to import services by some criteria (e.g. all services in default namespace with http ports`.

Otherwise to use the feature people would have to add annotations to try out the feature. I am looking for ways to make first experience "auto-magical"

Reply:
Maybe I wasn't clear enough in the RFD (I'll try to fix it), but my idea was also for the experience to be "auto-magical". On the first run we will import everything that matched config in the discovery service (namespaces and labels). We'll try our best to make it work, that's why by default TCP type is used, but also will use hints from Kubernetes `appProtocol`. Annotations are just an optional mechanism for finer control, if for example they don't want all 3 ports of an service to be turned into apps, as they only need one etc

Reply:
We can add some protocol detection logic to port selection based on the port name:
- `http`, `https`: automatically create an HTTP application
- `tcp`: automatically create a TCP app

For other cases, we can try `GET / HTTP1.1` and check if the response is HTTP - even if it's a 404 -  we mark it as HTTP to avoid having annotations. If we could not detect it, we fall back to TCP apps.

With this, it would cover most use cases in K8S. We can still allow users to select the protocol via annotation and automatic protocol detection can work with `dbs` as well.

Reply:
Can we consider an approach where we reduce the configuration using our Kubernetes Discover wizard while we offer an interactive way of visualizing the effect of the discovery config? 

Something like:
- Kubernetes Discover wizard automatically creates a token for `app,kube,discovery`
- Helm chart configures roles, dynamic app, and discovery resources.
- After the current form of Kubernetes Discover wizard, a new panel is shown to configure the automatic discovery of Kubernetes services (We can leverage the work that @marcoandredinis did for dynamic discovery configs [RFD 125](https://github.com/gravitational/teleport/pull/26613) and support K8S Service discovery)
- The panel allows users to edit, create and preview the impact of the discovery config. We can show the Kubernetes services that match the discovery criteria and their imported names - probably protocols too. With his data, they can adjust the discovery config.
- Once users are happy with the config, we create the dynamic discovery config and everything is magically imported into Teleport.

With an approach similar to this we can reduce the configuration effort,  allow users to preview the discovery config before creating it, improves security, and guides them when service annotations are required - we can even show them the kubectl command to annotate the service.

What do you think?

Reply:
We can figure out the exact Discover UX but yes, it should be enabled by default and not require extra tweaking on user's part. I think that if we add this to our Helm chart, we can default discovery configuration to wildcard and let users tweak it if needed like we're doing for other services.